### PR TITLE
SAI/Scripting: Add SMART_ACTION_BECOME_PERSONAL_CLONE_FOR_PLAYER

### PR DESCRIPTION
**Pull request by Seyden**
**Changes proposed:**

-  Add SMART_ACTION_BECOME_PERSONAL_CLONE_FOR_PLAYER

Tested with 2 clients:
.npc add 1141
.quest add 1271

+ smart script

```
-- Timed list 114100 smart ai
SET @ENTRY := 114100;
DELETE FROM `smart_scripts` WHERE `source_type` = 9 AND `entryOrGuid` = @ENTRY;
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(@ENTRY, 9, 0, 0, 0, 0, 100, 0, 0, 0, 0, 0, 149, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'After 0 seconds - Become Personal Clone for Player'),
(@ENTRY, 9, 1, 0, 0, 0, 100, 0, 0, 0, 0, 0, 83, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'After 0 seconds - Self: Remove npc flags QUESTGIVER'),
(@ENTRY, 9, 2, 0, 0, 0, 100, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'After 0 seconds - Self: Talk 0 to invoker'),
(@ENTRY, 9, 3, 0, 0, 0, 100, 0, 0, 0, 0, 0, 70, 120, 0, 0, 0, 0, 0, 15, 0, 100, 0, 0, 0, 0, 0, 'After 0 seconds - Any gameobject in 100 yd: Set respawn timer to 120 ms'),
(@ENTRY, 9, 4, 0, 0, 0, 100, 0, 5000, 5000, 0, 0, 66, 0, 0, 0, 0, 0, 0, 8, 0, 0, 0, 0, 0, 0, 1.02834, 'After 5 seconds - Self: Set orientation to 1.02834'),
(@ENTRY, 9, 5, 0, 0, 0, 100, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'After 0 seconds - Self: Talk 1 to invoker'),
(@ENTRY, 9, 6, 0, 0, 0, 100, 0, 6000, 6000, 0, 0, 100, 1, 0, 0, 0, 0, 0, 19, 3628, 0, 0, 0, 0, 0, 0, 'After 6 seconds - Send stored target storedTarget[1] to Closest alive creature Steven Lohan (3628) in 100 yards'),
(@ENTRY, 9, 7, 0, 0, 0, 100, 0, 0, 0, 0, 0, 45, 1, 1, 0, 0, 0, 0, 19, 3628, 0, 0, 0, 0, 0, 0, 'After 0 seconds - Closest alive creature Steven Lohan (3628) in 100 yards: Set creature data #1 to 1'),
(@ENTRY, 9, 8, 0, 0, 0, 100, 0, 4000, 4000, 0, 0, 82, 2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'After 4 seconds - Self: Add npc flags QUESTGIVER'),
(@ENTRY, 9, 9, 0, 0, 0, 100, 0, 0, 0, 0, 0, 66, 0, 0, 0, 0, 0, 0, 8, 0, 0, 0, 0, 0, 0, 0.558505, 'After 0 seconds - Self: Set orientation to 0.558505');
```

**Issues addressed:**

Closes #  (insert issue tracker number)

**Tests performed:**

(Does it build, tested in-game, etc.)

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] 
- [ ]

### CODE REVIEW

**Comment by Shauren** [comment]
Sql for testing is broken, missing action_type

**Comment by Seyden** [comment]
> Sql for testing is broken, missing action_type

Missing action type and action Param1 fixed

**Review by Shauren — CHANGES_REQUESTED** [review]

**Comment by BAndysc** [comment]
How about using targets instead of "GetLastInvoker()" which can work nondeterministic, e.g. in timed action list (other event can override and "last" invoker is different than user expects it to be). Stored targets solves this

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -2522,6 +2522,24 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
...
+                if (!IsPlayer(target))
+                    continue;
+
+                ObjectGuid privateObjectOwner = target->GetGUID();
+                if (Creature* summon = GetBaseObject()->SummonPersonalClone((TempSummonType)e.action.becomePersonalClone.type, e.action.becomePersonalClone.duration, 0, 0, privateObjectOwner))
+                {
+                    if (IsSmart(summon))
+                        ENSURE_AI(SmartAI, summon->AI())->SetTimedActionList(e, e.entryOrGuid, GetLastInvoker(), e.event_id + 1);
+
+                    Trinity::Containers::EraseIf(mTimedActionList, [e](SmartScriptHolder const& script) { return script.event_id >= e.event_id; });
```
Comment (Shauren) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:2538]: Invalidates `e` (accessed later at the end of `ProcessAction`), erase `>`, not `>=`

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 2]
```diff
@@ -2522,6 +2522,24 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
...
+            for (WorldObject* target : targets)
+            {
+                if (!IsPlayer(target))
+                    continue;
+
+                ObjectGuid privateObjectOwner = target->GetGUID();
+                if (Creature* summon = GetBaseObject()->SummonPersonalClone((TempSummonType)e.action.becomePersonalClone.type, e.action.becomePersonalClone.duration, 0, 0, privateObjectOwner))
+                {
+                    if (IsSmart(summon))
+                        ENSURE_AI(SmartAI, summon->AI())->SetTimedActionList(e, e.entryOrGuid, GetLastInvoker(), e.event_id + 1);
```
Comment (Shauren) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:2536]:
```suggestion
                        ENSURE_AI(SmartAI, summon->AI())->SetTimedActionList(e, e.entryOrGuid, target, e.event_id + 1);
```

File: src/server/game/Entities/Object/Object.cpp [hunk 3]
```diff
@@ -1840,6 +1840,20 @@ Scenario* WorldObject::GetScenario() const
     return nullptr;
 }
 
+TempSummon* WorldObject::SummonPersonalClone(TempSummonType despawnType, uint32 despawnTime, uint32 vehId, uint32 spellId, ObjectGuid privateObjectOwner)
+{
+    if (Map* map = FindMap())
+    {
+        if (TempSummon* summon = map->SummonCreature(this->GetEntry(), *this, nullptr, despawnTime, this, spellId, vehId, privateObjectOwner))
```
Comment (Shauren) [inline src/server/game/Entities/Object/Object.cpp:1847]:
```suggestion
        if (TempSummon* summon = map->SummonCreature(GetEntry(), GetPosition(), nullptr, despawnTime, this, spellId, vehId, privateObjectOwner))
```

**Review by Shauren — CHANGES_REQUESTED** [review]

**Pull request merged**